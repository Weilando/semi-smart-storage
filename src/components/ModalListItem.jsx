import React from 'react';
import PropTypes from 'prop-types';
import '../styles/Modal.css';

function ModalListItem(props) {
  ModalListItem.propTypes = {
    obj: PropTypes.shape(
      {id: PropTypes.number.isRequired, name: PropTypes.string.isRequired}
    ).isRequired,
    editAction: PropTypes.func.isRequired,
  }

  return (
    <li className="Modal">
      <button className="Block"><p>{props.obj.name}</p></button>
      <button className="Round" onClick={() => props.editAction(props.obj)}>
        i
      </button>
    </li>
  );
}

export default ModalListItem;
