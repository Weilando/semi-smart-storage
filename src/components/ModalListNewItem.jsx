import React from 'react';
import PropTypes from 'prop-types';
import '../styles/Modal.css';

class ModalListNewItem extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: '',
    }

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  static propTypes = {
    addAction: PropTypes.func.isRequired,
  }

  handleChange(event) {
    this.setState({name: event.target.value});
  }

  handleSubmit(event) {
    event.preventDefault();
    this.props.addAction(this.state.name);
    this.setState({name: ''});
  }

  render() {
    return (
      <li className="Modal">
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            className="ModalList"
            value={this.state.name}
            onChange={this.handleChange}
          />
          <input
            type="submit"
            className="ModalList"
            value="Add"
            disabled={this.state.name === ''}
          />
        </form>
      </li>
    );
  }

}

export default ModalListNewItem;
