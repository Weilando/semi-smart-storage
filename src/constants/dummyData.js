const DummyItemList = [
  {id: 0, name: "Sparkling water"},
  {id: 1, name: "Apple juice"},
  {id: 2, name: "Craft beer"},
  {id: 3, name: "Milk"},
  {id: 5, name: "Coke"}
];

const DummyItemListShort = [
  {id: 0, name: "Sparkling water"},
  {id: 1, name: "Apple juice"},
  {id: 3, name: "Milk"}
];

const DummyStorageContent = [
  {id: 0, item: "Milk", unit: "1L", quantity: 2.7},
  {id: 1, item: "Coke", unit: "0.33L", quantity: 5},
  {id: 2, item: "Apple juice", unit: "0.5L", quantity: 2},
  {id: 4, item: "Craft Beer", unit: "0.33L", quantity: 6},
  {id: 5, item: "Sparkling water", unit: "0.75L", quantity: 21}
];

const DummyStorageContentShort = [
  {id: 0, item: "Milk", unit: "1L", quantity: 2.7},
  {id: 2, item: "Sparkling water", unit: "0.75L", quantity: 21}
];

const DummyStorageList = [
  {id: 0, name: 'Fridge'},
  {id: 1, name: 'Basement'},
  {id: 3, name: 'Garage'},
];

const DummyStorageListShort = [
  {id: 0, name: 'Basement'},
  {id: 3, name: 'Fridge'},
];

const DummyUnitList = [
  {id: 0, name: "0.33L"},
  {id: 1, name: "0.5L"},
  {id: 2, name: "0.75L"},
  {id: 3, name: "1.0L"}
];

const DummyUnitListShort = [
  {id: 0, name: "0.33L"},
  {id: 2, name: "0.75L"},
  {id: 3, name: "1.0L"}
];

export { DummyItemList, DummyStorageContent, DummyStorageList, DummyUnitList };
export { DummyItemListShort, DummyStorageContentShort, DummyStorageListShort, DummyUnitListShort };
