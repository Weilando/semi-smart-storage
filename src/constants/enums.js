// api-operations
const AddMode = {
  ITEM_LIST: 0, STORAGE_LIST: 1, STORAGE_CONTENT: 2, UNIT_LIST: 3
}

const DeleteMode = {
  ITEM_LIST: 0, STORAGE_LIST: 1, STORAGE_CONTENT: 2, UNIT_LIST: 3
}

const FetchMode = {
  ITEM_LIST: 0, STORAGE_LIST: 1, STORAGE_CONTENT: 2, UNIT_LIST: 3
}

const UpdateMode = {
  ITEM_LIST: 0,
  STORAGE_LIST: 1,
  STORAGE_CONTENT_DECREMENT: 2,
  STORAGE_CONTENT_INCREMENT: 3,
  STORAGE_CONTENT_EDIT: 4,
  UNIT_LIST: 5
}

// modal-modes
const ModalMode = {
  ITEM: "item",
  STORAGE: "storage",
  UNIT: "unit"
}

export { AddMode, DeleteMode, FetchMode, UpdateMode, ModalMode };
