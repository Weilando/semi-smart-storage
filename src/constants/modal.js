let ModalMode = {
  ITEM: "item",
  STORAGE: "storage",
  UNIT: "unit"
}

export { ModalMode }
