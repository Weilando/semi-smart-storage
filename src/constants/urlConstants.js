const BASEURL = "http://raspberrypi/api/";
const ADD = "add.php";
const DELETE = "delete.php";
const GET = "get.php";
const UPDATE = "update.php";

export { BASEURL };
export { ADD, DELETE, GET, UPDATE };
