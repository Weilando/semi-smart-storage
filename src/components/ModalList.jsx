import React from 'react';
import PropTypes from 'prop-types';
import ModalListItem from './ModalListItem';
import ModalListNewItem from './ModalListNewItem';
import { ModalMode } from '../constants/modal';
import '../styles/Modal.css';

function unpackList(list, editAction) {
  const keys = [...Array(list.length).keys()]; // Array with keys from 0 to entries.length

  return keys.map((currKey) =>
    <ModalListItem
      key={list[currKey].id}
      obj={list[currKey]}
      editAction={editAction}
    />
  );
}

function ModalList(props) {
  ModalList.propTypes = {
    show: PropTypes.bool.isRequired,
    list: PropTypes.arrayOf(
      PropTypes.shape(
        {id: PropTypes.number.isRequired, name: PropTypes.string.isRequired}
      )
    ).isRequired,
    mode: PropTypes.oneOf([ModalMode.ITEM, ModalMode.UNIT]),
    addAction: PropTypes.func.isRequired,
    closeAction: PropTypes.func.isRequired,
    editAction: PropTypes.func.isRequired,
  }

  const list = unpackList(props.list, props.editAction);

  return(
    <div className={'modal '.concat(props.show ? 'show' : 'hide')}>
      <div className="modal-main">
        <span className="close" onClick={props.closeAction}>&times;</span>

        <h1>Settings for {props.mode}</h1>
        <p>Click the edit-button next to the {props.mode} you want to edit and perform your changes.</p>

        <ul className="Modal">
          {list}
          <ModalListNewItem addAction={props.addAction}/>
        </ul>
      </div>
    </div>
  );
}

export default ModalList;
