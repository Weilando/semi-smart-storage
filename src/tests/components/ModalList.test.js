import React from 'react';
import renderer from 'react-test-renderer';
import ModalList from '../../components/ModalList';
import ModalListItem from '../../components/ModalListItem';
import ModalListNewItem from '../../components/ModalListNewItem';
import { ModalMode } from '../../constants/enums';
import { DummyItemListShort, DummyUnitListShort } from '../../constants/dummyData';

describe('snapshot tests for ModalList component', () => {
  it('should render visible ModalList component for items correctly', () => {
    const addAction = jest.fn();
    const closeAction = jest.fn();
    const editAction = jest.fn();

    const tree = renderer.create(
      <ModalList
        show={true}
        list={DummyItemListShort}
        mode={ModalMode.ITEM}
        addAction={addAction}
        closeAction={closeAction}
        editAction={editAction}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  })

  it('should render hidden ModalEdit component for units correctly', () => {
    const addAction = jest.fn();
    const closeAction = jest.fn();
    const editAction = jest.fn();

    const tree = renderer.create(
      <ModalList
        show={false}
        list={DummyUnitListShort}
        mode={ModalMode.ITEM}
        addAction={addAction}
        closeAction={closeAction}
        editAction={editAction}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  })
});

describe('snapshot tests for ModalListItem component', () => {
  it('should render ModalListItem component correctly', () => {
    const obj = {id: 0, name: "Sparkling water"}
    const editAction = jest.fn();

    const tree = renderer.create(
      <ModalListItem
        obj={obj}
        editAction={editAction}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  })
});

describe('snapshot tests for ModalListNewItem component', () => {
  it('should render ModalListNewItem component correctly', () => {
    const addAction = jest.fn();

    const tree = renderer.create(
      <ModalListNewItem
        addAction={addAction}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  })
});
