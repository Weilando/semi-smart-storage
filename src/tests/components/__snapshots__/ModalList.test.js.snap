// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests for ModalList component should render hidden ModalEdit component for units correctly 1`] = `
<div
  className="modal hide"
>
  <div
    className="modal-main"
  >
    <span
      className="close"
      onClick={[MockFunction]}
    >
      ×
    </span>
    <h1>
      Settings for 
      item
    </h1>
    <p>
      Click the edit-button next to the 
      item
       you want to edit and perform your changes.
    </p>
    <ul
      className="Modal"
    >
      <li
        className="Modal"
      >
        <button
          className="Block"
        >
          <p>
            0.33L
          </p>
        </button>
        <button
          className="Round"
          onClick={[Function]}
        >
          i
        </button>
      </li>
      <li
        className="Modal"
      >
        <button
          className="Block"
        >
          <p>
            0.75L
          </p>
        </button>
        <button
          className="Round"
          onClick={[Function]}
        >
          i
        </button>
      </li>
      <li
        className="Modal"
      >
        <button
          className="Block"
        >
          <p>
            1.0L
          </p>
        </button>
        <button
          className="Round"
          onClick={[Function]}
        >
          i
        </button>
      </li>
      <li
        className="Modal"
      >
        <form
          onSubmit={[Function]}
        >
          <input
            className="ModalList"
            onChange={[Function]}
            type="text"
            value=""
          />
          <input
            className="ModalList"
            disabled={true}
            type="submit"
            value="Add"
          />
        </form>
      </li>
    </ul>
  </div>
</div>
`;

exports[`snapshot tests for ModalList component should render visible ModalList component for items correctly 1`] = `
<div
  className="modal show"
>
  <div
    className="modal-main"
  >
    <span
      className="close"
      onClick={[MockFunction]}
    >
      ×
    </span>
    <h1>
      Settings for 
      item
    </h1>
    <p>
      Click the edit-button next to the 
      item
       you want to edit and perform your changes.
    </p>
    <ul
      className="Modal"
    >
      <li
        className="Modal"
      >
        <button
          className="Block"
        >
          <p>
            Sparkling water
          </p>
        </button>
        <button
          className="Round"
          onClick={[Function]}
        >
          i
        </button>
      </li>
      <li
        className="Modal"
      >
        <button
          className="Block"
        >
          <p>
            Apple juice
          </p>
        </button>
        <button
          className="Round"
          onClick={[Function]}
        >
          i
        </button>
      </li>
      <li
        className="Modal"
      >
        <button
          className="Block"
        >
          <p>
            Milk
          </p>
        </button>
        <button
          className="Round"
          onClick={[Function]}
        >
          i
        </button>
      </li>
      <li
        className="Modal"
      >
        <form
          onSubmit={[Function]}
        >
          <input
            className="ModalList"
            onChange={[Function]}
            type="text"
            value=""
          />
          <input
            className="ModalList"
            disabled={true}
            type="submit"
            value="Add"
          />
        </form>
      </li>
    </ul>
  </div>
</div>
`;

exports[`snapshot tests for ModalListItem component should render ModalListItem component correctly 1`] = `
<li
  className="Modal"
>
  <button
    className="Block"
  >
    <p>
      Sparkling water
    </p>
  </button>
  <button
    className="Round"
    onClick={[Function]}
  >
    i
  </button>
</li>
`;

exports[`snapshot tests for ModalListNewItem component should render ModalListNewItem component correctly 1`] = `
<li
  className="Modal"
>
  <form
    onSubmit={[Function]}
  >
    <input
      className="ModalList"
      onChange={[Function]}
      type="text"
      value=""
    />
    <input
      className="ModalList"
      disabled={true}
      type="submit"
      value="Add"
    />
  </form>
</li>
`;
